File: solver.py
#!/usr/bin/env python3

import re
import json
from pathlib import Path


class UserInputError(Exception):
    pass


class Solved(Exception):
    pass


def _reorder_by_letter(words):
    output = []
    for i in "qjzxvkwyfbghmpduclsntoirae":
        for j in words[::-1]:
            if i in j and j not in output:
                output.append(j)
    return output


def _reorder_by_occurrences(words):
    output = []
    for i in words:
        for j in range(5):
            if i.count(i[j]) != 1:
                output.insert(0, i)
                break
        else:
            output.append(i)
    return output


def reorder(words, board):
    return _reorder_by_occurrences(_reorder_by_letter(words)), board


def make_regex(board):
    regex = ""
    for i in range(5):
        if board[i][2] != []:
            regex += board[i][2][0]
            continue
        regex += "[^"
        for j in board[i][0]:
            regex += j
        regex += "]"
    return regex


def make_second_regex(board):
    output = []
    for i in range(5):
        for j in board[i][1]:
            if j not in output:
                output.append(j)
    return [f".*{i}.*" for i in output]


def zero_maybes(board):
    for i in range(5):
        board[i][1] = []
    return board


def remove_nos_from_maybes(board):
    for i in range(5):
        letter = board[i]
        for j in range(len(letter[0])):
            if letter[0][j] in letter[1]:
                letter[1].remove(letter[0][j])
    return board


def guard_inputs(word, status):
    if len(word) != 5 or len(status) != 5 or not status.isdigit() or not word.isalpha():
        raise UserInputError("Are you sure you entered the word and status correctly?")
    for i in status:
        if i not in "012":
            raise UserInputError("Status must be just (0, 1, 2)")
    if status == "22222":
        raise Solved("You solved it!")
    return word, status


def add_to_board(word, status, board):
    for i in range(5):
        if status[i] == "0":
            for letter in board:
                if word[i] not in letter[0]:
                    letter[0].append(word[i])
        elif status[i] == "1":
            for j in range(5):
                if j == i:
                    if word[i] not in board[i][0]:
                        board[i][0].append(word[i])
                else:
                    if word[i] not in board[j][1]:
                        board[j][1].append(word[i])
        elif status[i] == "2":
            if word[i] not in board[i][2]:
                board[i][2].append(word[i])
    return board


def get_remaining(words, board):
    remaining_words = []
    first_regex = re.compile(make_regex(board))
    for i in words:
        if first_regex.match(i):
            remaining_words.append(i)
    for pattern in make_second_regex(board):
        temp = []
        second_regex = re.compile(pattern)
        for word in remaining_words:
            if second_regex.match(word):
                temp.append(word)
        remaining_words = temp.copy()
    return remaining_words, board


def init_board():
    return [
        [[] for _ in range(3)] for _ in range(5)
    ]  # [no, yes, maybe] for letter in range(5)


def init_words_list():
    with open(f"{Path(__file__).parent}/words.json", "r") as f:
        return json.load(f)


def get_words_board_wrapper(word, status, board=init_board()):
    """wraps many functions defined to make running much simpler

    Args:
        word (str): a word to test
        status (str): a string representation of an input status (consists of 0, 1, 2)
        board (list of lists - board object): this is the current status of the entire board so far. If no board is provided, it will default to a blank board

    Returns:
        words (list): ordered list of possible remaining words given the inputs
        board: the input board so this function can be called in a loop
    """
    return reorder(
        *get_remaining(
            init_words_list(),
            remove_nos_from_maybes(
                add_to_board(
                    *guard_inputs(word.lower(), status),
                    zero_maybes(board),
                ),
            ),
        )
    )


if __name__ == "__main__":
    print("Status: 0 for gray, 1 for yellow, 2 for green\nSolved: 22222")
    count = 1
    last_word = ""
    board = init_board()

    while True:
        try:
            output, board = get_words_board_wrapper(
                input("Word: ") or last_word, input("Status: "), board
            )
        except UserInputError as e:
            print(e)
            continue
        except Solved as e:
            print(e)
            break
        if len(output) == 0:
            print("No words left")
            break
        last_word = output[-1] if output else ""
        print("\n".join(output[:-1]), f"\n{last_word}", len(output), sep="\n")
        count += 1
    print(f"{count} words entered")

File: solver.js
#!/usr/bin/env node

const fs = require('fs');
const readline = require('readline-sync');

const words = JSON.parse(fs.readFileSync("words.json"));

let board = [[[], [], []], [[], [], []], [[], [], []], [[], [], []], [[], [], []]];

class UserInputError extends Error {
    constructor(message) {
        super(message);
        this.name = 'UserInputError';
    }
}

class Solved extends Error {
    constructor(message) {
        super(message);
        this.name = 'Solved';
    }
}

function reorder(words) {
    let test_1 = [];
    for (i of "qjzxvkwyfbghmpduclsntoirae") {
        for (j of words.reverse()) {
            if (j.includes(i)) {
                if (!test_1.includes(j)) {
                    test_1.push(j);
                }
            }
        }
    }
    let test_2 = [];
    for (i of test_1) {
        var flag = false;
        for (j = 0; j < 5; j++) {
            for (count = -1, index = -2; index != -1; count++, index = i.indexOf(i[j], index + 1));
            if (count != 1) {
                test_2.unshift(i);
                flag = true;
                break;
            }
        }
        if (!flag) {
            test_2.push(i);
        }
    }
    return test_2;
}

function make_regex(board) {
    let regex = "";
    for (var i = 0; i < 5; i++) {
        if (board[i][2].length != 0) {
            regex += board[i][2][0];
        } else {
            regex += "[^";
            for (j of board[i][0]) {
                regex += j;
            }
            regex += "]";
        }
    }
    return regex;
}

function make_second_regex(board) {
    let output = [];
    for (i = 0; i < 5; i++) {
        for (j of board[i][1]) {
            if (!(j in output)) {
                output.push(`.*${j}.*`);
            }
        }
    }
    return output;
}

function is_alpha(string) {
    return /^[a-zA-Z]+$/.test(string);
}

function is_digit(string) {
    return /^[0-9]+$/.test(string);
}

function get_remaining(word, status) {
    if (
        word.length != 5 ||
        status.length != 5 ||
        !is_alpha(word) ||
        !is_digit(status)
    ) {
        throw new UserInputError("Invalid input");
    }
    if (status == "22222") {
        throw new Solved("You solved it!");
    }
    for (i = 0; i < 5; i++) {
        board[i][1] = [];
    }
    for (i = 0; i < 5; i++) {
        if (status[i] == "0") {
            for (j of board) {
                if (!(word[i] in j[0])) {
                    j[0].push(word[i]);
                }
            }
        } else if (status[i] == "1") {
            for (j = 0; j < 5; j++) {
                if (j == i) {
                    if (!(word[i] in board[i][0])) {
                        board[i][0].push(word[i]);
                    }
                } else {
                    if (!(word[i] in board[j][1])) {
                        board[j][1].push(word[i]);
                    }
                }
            }
        } else if (status[i] == "2") {
            if (!(word[i] in board[i][2])) {
                board[i][2].push(word[i]);
            }
        }
    }
    for (i = 0; i < 5; i++) {
        for (j = 0; j < board[i][0].length; j++) {
            if (board[i][0][j] in board[i][1]) {
                board[i][1].remove(board[i][0][j]);
            }
        }
    }
    let remaining_words = [];
    for (i of words) {
        if (new RegExp(make_regex(board)).test(i)) {
            remaining_words.push(i);
        }
    }
    for (pattern of make_second_regex(board)) {
        let temp = [];
        for (i of remaining_words) {
            if (i.match(new RegExp(pattern)) != null) {
                temp.push(i);
            }
        }
        remaining_words = Array.from(temp);
    }
    return remaining_words;
}

function input(string) {
    return readline.question(string);
}

function main() {
    console.log("Status: 0 for gray, 1 for yellow, 2 for green\nSolved: 22222");
    let count = 1;
    let last_word = "";
    let output = [];
    while (true) {
        try {
            output = reorder(get_remaining(input("Word: ") || last_word, input("Status: ")));
        } catch (e) {
            if (e instanceof UserInputError) {
                console.log(e.message);
                continue;
            } else if (e instanceof Solved) {
                console.log(e.message);
                break;
            } else {
                throw e;
            }
        }
        if (output.length == 0) {
            console.log("No words left");
            break;
        }
        last_word = output.length > 0 ? output[output.length-1] : "";
        output.slice(0, -1).forEach(element => {
            console.log(element);
        });
        console.log();
        console.log(last_word);
        console.log(output.length);
        count++;
    }
    console.log(`${count} words entered`);
}

main()